pallets/metamask/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+
+    #[pallet::config(with_default)]
+    pub trait Config: frame_system::Config + pallet_evm_chain_id::Config {
+        type Sender: Parameter + Into<sp_core::H160> + Into<Self::AccountId>;
+        type Nonce: Parameter + Into<sp_core::U256> + PartialEq;
+
+        #[pallet::no_default_bounds]
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        #[pallet::no_default_bounds]
+        type RuntimeCall: Parameter
+            + UnfilteredDispatchable<RuntimeOrigin = Self::RuntimeOrigin>
+            + GetDispatchInfo;
+
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight({
+            let dispatch_info = call.get_dispatch_info();
+            (dispatch_info.weight, dispatch_info.class)
+        })]
+        pub fn signed_call(
+            _: OriginFor<T>,
+            sender: <T as Config>::Sender,
+            nonce: <T as Config>::Nonce, // sender nonce, part of signed data to prevent replay attack
+            signature: Vec<u8>,          // 'eth_signTypedData_v4' result
+            call: Box<<T as Config>::RuntimeCall>,
+        ) -> DispatchResultWithPostInfo {
+            // TODO: mess with the nonce: synch check with `validate_unsigned` and increment
+            let chain_id = pallet_evm_chain_id::Pallet::<T>::get();
+
+            {
+                let sender: sp_core::H160 = sender.clone().into();
+                let call = call.encode(); // SCALE
+                let call = bs58::encode(call).as_cb58(None).into_vec();
+                let nonce: sp_core::U256 = nonce.into();
+
+                let domain = Domain {
+                    name: b"ATLA".into(),
+                    version: b"1".into(),
+                    chain_id: chain_id.into(),
+                    verifying_contract: sp_core::H160::zero(),
+                };
+
+                let payload = Payload { sender, nonce, call };
+                let typed_data = TypedData::new(domain, payload);
+
+                let hash = typed_data.message_hash();
+
+                let signature =
+                    parse_signature(&signature).map_err(|_| Error::<T>::BadSignatureFormat)?;
+
+                let origin = recover_signer_address(signature, hash)
+                    .map_err(|_| Error::<T>::EcdsaRecoverErr)?;
+
+                frame_support::ensure!(sender == origin, Error::<T>::SignerMismath);
+            }
+
+            let signer: <T as frame_system::Config>::AccountId = sender.into();
+            let result = call.dispatch_bypass_filter(RawOrigin::Signed(signer.clone()).into());
+            Self::deposit_event(Event::Authorized {
+                signed_by: signer,
+                call_result: result.map(|_| ()).map_err(|e| e.error),
+            });
+
+            Ok(Pays::No.into()) // TODO: take the fee from account
```

Comment:
now I'm really not sure about safety here. this call is unsigned and feeless. nothing prevents attacking us here. if I didn't miss something. instead we could add restriction for who can call this. and then we can call it using sudo account, for example. this won't be much a problem for the frontends, as they have a dedicated backend, which is partially a gateway to the node's rpc. the backend could provide the method to call this with data, while using sudo call to perform actual dispatch.

Reply:
ah, I forgot that this call dispatches a call... anyway, feels scary. we need to check it.

Reply:
Yes, we need to take call fee. I'm just not sure if this is possible take call fee without signing the extrinsic itself. So it requires some attention in the future.

Everything related to security is the question how secp256k1 _recover from signature_ function works. If it is safe in from a cryptographic point of view, then we are safe too. But we can add some additional restrictions if necessary.
__TBD__...